File: Patterns/Proxy Pattern/CSharp/Subject/Proxy.cs
namespace Subject
{
    public class Proxy : Image
    {
        private Image? image;
        private string path;

        public Proxy(string path) 
        {
            image = null;
            this.path = path;
        }

        public void Render() 
        {
            if (image == null) image = new Subject(path);
            image.Render();
        }
    }
}

File: Patterns/Proxy Pattern/Cpp/Subject/Proxy.cpp
#pragma once

#include <string>
#include "Subject.cpp"

using namespace std;

class Proxy : public Image {
private:
    Image* image;
    string path;

public:
    Proxy(string path) : image(nullptr), path(path) {}

    ~Proxy() override {
        if (image != nullptr) delete image;
    }

    void render() override {
        if (image == nullptr) image = new Subject(path);
        image->render();
    }
};